File: exercism/c/c/pangram/src/pangram.c
#include "pangram.h"

#include <ctype.h>

bool is_pangram(const char *sentence) {
	if (! sentence)
		return false;
	
	int mask = 0;

	for (int i = 0; sentence[i] != '\0'; i++) {
		char c = sentence[i];

		if (!isalpha(c))
			continue;
		if (isupper(c))
			c += 32;
		
		mask |= (1 << (c - 97));
	}

	// check at the end if 26 bits are set.
	for (int i = 0; i < 26; i++) {
		if ((mask & (1 << i)) == 0)
			return false;
	}

	return true;
}

File: exercism/go/go/pangram/pangram.go
// Package pangram detects if a sentence is a pangram 
package pangram

import "unicode"

// IsPangram returns true if a sentence is a pangram
// else returns false
func IsPangram(input string) bool {
	if input == "" {
		return false
	}
	
	var mask uint32 = 0

	for _, c := range input {
		if !unicode.IsLetter(c) {
			continue
		}
	
		// 0011 1111 1111 1111 1111 1111 1111 == 2FFFFFF
		c = unicode.ToLower(c)
		placesToShift := uint32((int(c) - 97))
		mask |= (1 << placesToShift)
	}

	if mask & 0x3FFFFFF == 0x3FFFFFF {
		return true
	}
	return false
}
